source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -492,42 +498,58 @@ HeaderValidator::HeaderEntryValidationResult HeaderValidator::validateGenericReq
-        const auto& header_name = header_entry.key();
-        const auto& header_value = header_entry.value();
-
-        auto entry_name_result = validateGenericHeaderName(header_name);
-        if (entry_name_result.action() == HeaderEntryValidationResult::Action::DropHeader) {
-          // drop the header, continue processing the request
-          drop_headers.push_back(header_name.getStringView());
-        } else if (!entry_name_result) {
-          reject_details = static_cast<std::string>(entry_name_result.details());
-        } else {
-          auto entry_value_result = validateGenericHeaderValue(header_value);
-          if (!entry_value_result) {
-            reject_details = static_cast<std::string>(entry_value_result.details());
-          }
-        }
-
-        return reject_details.empty() ? ::Envoy::Http::HeaderMap::Iterate::Continue
-                                      : ::Envoy::Http::HeaderMap::Iterate::Break;
-      });
+  trailers.iterate([this, &reject_details](const ::Envoy::Http::HeaderEntry& header_entry)
+                       -> ::Envoy::Http::HeaderMap::Iterate {
+    const auto& header_name = header_entry.key();
+    const auto& header_value = header_entry.value();
+
+    auto entry_name_result = validateGenericHeaderName(header_name);
+    if (!entry_name_result.ok()) {
+      reject_details = static_cast<std::string>(entry_name_result.details());
+    } else {
+      auto entry_value_result = validateGenericHeaderValue(header_value);
+      if (!entry_value_result) {
+        reject_details = static_cast<std::string>(entry_value_result.details());
+      }
+    }
+
+    return reject_details.empty() ? ::Envoy::Http::HeaderMap::Iterate::Continue
+                                  : ::Envoy::Http::HeaderMap::Iterate::Break;
+  });
 
   if (!reject_details.empty()) {
-    return {TrailerValidationResult::Action::Reject, reject_details};
+    return {ValidationResult::Action::Reject, reject_details};
   }
 
-  for (auto& name : drop_headers) {
-    trailers.remove(::Envoy::Http::LowerCaseString(name));
+  return ValidationResult::success();
+}
+
+void HeaderValidator::sanitizeHeadersWithUnderscores(::Envoy::Http::HeaderMap& header_map) {
+  const auto& underscore_action = config_.headers_with_underscores_action();
+  if (underscore_action == HeaderValidatorConfig::ALLOW) {
+    return;
   }
 
-  return TrailerValidationResult::success();
+  std::vector<absl::string_view> drop_headers;
+  header_map.iterate([&drop_headers](const ::Envoy::Http::HeaderEntry& header_entry)
+                         -> ::Envoy::Http::HeaderMap::Iterate {
+    const absl::string_view header_name = header_entry.key().getStringView();
+    if (absl::StrContains(header_name, '_')) {
+      drop_headers.push_back(header_name);
+    }
+
+    return ::Envoy::Http::HeaderMap::Iterate::Continue;
+  });
+
+  // The reject case is handled in the validate... methods
```

Comment:
validate... ?

Reply:
I have moved the explanation to the header comment.